cli/lsp/logging.rs

```diff
@@ -1,13 +1,90 @@
 // Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
 
+use chrono::DateTime;
+use chrono::Utc;
+use deno_core::parking_lot::Mutex;
+use std::fs;
+use std::io::prelude::*;
+use std::path::PathBuf;
 use std::sync::atomic::AtomicBool;
 use std::sync::atomic::AtomicUsize;
 use std::sync::atomic::Ordering;
+use std::thread;
+use std::time::SystemTime;
 
 static LSP_DEBUG_FLAG: AtomicBool = AtomicBool::new(false);
 static LSP_LOG_LEVEL: AtomicUsize = AtomicUsize::new(log::Level::Info as usize);
 static LSP_WARN_LEVEL: AtomicUsize =
   AtomicUsize::new(log::Level::Warn as usize);
+static LOG_FILE: LogFile = LogFile {
+  enabled: AtomicBool::new(true),
+  path: Mutex::new(None),
+  buffer: Mutex::new(String::new()),
+};
+
+pub struct LogFile {
+  enabled: AtomicBool,
+  path: Mutex<Option<PathBuf>>,
+  buffer: Mutex<String>,
+}
+
+impl LogFile {
+  pub fn write_line(&self, s: &str) {
+    if LOG_FILE.enabled.load(Ordering::Relaxed) {
+      let mut buffer = self.buffer.lock();
+      buffer.push_str(s);
+      buffer.push('\n');
+    }
+  }
+
+  fn commit(&self) {
+    let unbuffered = {
+      let mut buffer = self.buffer.lock();
+      if buffer.is_empty() {
+        return;
+      }
+      let unbuffered = buffer.clone();
+      buffer.clear();
+      unbuffered
```

Comment:
Maybe? (not sure if compiles, but basically swapping the value out of the buffer instead of cloning and clearing)

```suggestion
      std::mem::take(&mut buffer)
```

Reply:
Good point. Clone and clear sounds good. Maybe add a comment about that as to why this is done?